lib/web/kube.go

```diff
@@ -0,0 +1,326 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package web
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gorilla/websocket"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	oteltrace "go.opentelemetry.io/otel/trace"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/remotecommand"
+
+	clientproto "github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/constants"
+	mfav1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/mfa/v1"
+	"github.com/gravitational/teleport/api/mfa"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/session"
+)
+
+// podHandler connects Kube exec session and web-based terminal via a websocket.
+type podHandler struct {
+	teleportCluster     string
+	configTLSServerName string
+	configServerAddr    string
+	req                 PodExecRequest
+	sess                session.Session
+	sctx                *SessionContext
+	ws                  *websocket.Conn
+	keepAliveInterval   time.Duration
+	log                 *logrus.Entry
+	userClient          auth.ClientI
+	localAccessPoint    localAccessPoint
+
+	// closedByClient indicates if the websocket connection was closed by the
+	// user (closing the browser tab, exiting the session, etc).
+	closedByClient atomic.Bool
+}
+
+// PodExecRequest describes a request to create a web-based terminal
+// to exec into a pod.
+type PodExecRequest struct {
+	// Namespace is the namespace of the target pod
+	Namespace string `json:"namespace"`
+	// Pod is the target pod to connect to.
+	Pod string `json:"pod"`
+	// Container is a container within the target pod to connect to, optional.
+	Container string `json:"container"`
+	// Command is the command to run at the target pod.
+	Command string `json:"command"`
+	// KubeCluster specifies what Kubernetes cluster to connect to.
+	KubeCluster string `json:"kube_cluster"`
+	// IsInteractive specifies whether exec request should have interactive TTY.
+	IsInteractive bool `json:"is_interactive"`
+	// Term is the initial PTY size.
+	Term session.TerminalParams `json:"term"`
+}
+
+// ServeHTTP sends session metadata to web UI to signal beginning of the session, then
+// handles Kube exec request and connects it to web based terminal input/output.
+func (p *podHandler) ServeHTTP(_ http.ResponseWriter, r *http.Request) {
```

Comment:
Is it possible to add test coverage to this handler? You can use kube integration tests if you want to test against a real kube cluster or use our mock kube api in lib/kube/proxy